### PR TITLE
Report participant scores when running `credrank`

### PR DESCRIPTION
**Pull request by decentralion**
This commit adds functional APIs for retrieving participants and their
Cred scores from the Cred Graph.

As of #2279, we no longer embed participant (scoring) nodes in the
Markov Process Graph. However, this means the CredGraph API for
retrieving user scores was rendered non-functional.

This commit fixes this, while munging several changes together:
- Users have been renamed to participants, for consistency
(organizations and projects are also participants)
- We track Participants rather than scoring addresses, so that we have
descriptions available for participants (since the nodes are no longer
in the graph, we need to store their descriptions separately)
- The MarkovProcessGraph now reports all the epoch boundaries, which
makes it feasible to enumerate all the payout edges
- We expose logic for constructing a payout edge address given the epoch
time boundary and the epoch owner
- We update the CredGraph so that instead of the (non-functional)
`scoringNodes` api, it has a `participants` iterator which returns
participants along with their total Cred and their Cred over Time.

Finally, I've updated the `credrank` command so that it prints a table
of the top-10 participants and their total Cred. This allows us to
replace the "run this complicated jq command" test plan with simply
running `credrank` and observing the results. Naturally, once we have
better test infrastructure in place, we will remove this logging.

Test plan: Run `credrank` in the SourceCred cred instance, and observe:

| Description | Cred |
| --- | --- |
| github/decentralion | 5117 |
| discord/decentralion#1337 | 4927 |
| discourse/decentralion | 3807 |
| github/wchargin | 3148 |
| discord/LBS#6541 | 2773 |
| discord/s_ben#7872 | 2327 |
| discourse/LB | 1773 |
| discourse/@s_ben | 1773 |
| discord/Beanow#5887 | 1645 |
| discord/hammadj#6801 | 1393 |

Note to reviewer: This could have been split into smaller / more atomic commits, but given that the code is untested and generally in a "high-heat" early stage of simulated annealing, munging this together made sense, and made for a much easier test plan. Let me know if you have pushback!

### CODE REVIEW

File: src/core/markovProcessGraph.js [hunk 1]
```diff
@@ -568,11 +594,15 @@ export class MarkovProcessGraph {
       });
     }
 
-    return new MarkovProcessGraph(_nodes, _edges, _scoringAddresses);
+    return new MarkovProcessGraph(_nodes, _edges, participants, timeBoundaries);
+  }
+
+  epochBoundaries(): $ReadOnlyArray<number> {
+    return this._epochBoundaries;
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core/markovProcessGraph.js:598]: used deepFreeze instead

File: src/cli/credrank.js [hunk 2]
```diff
@@ -80,18 +88,34 @@ const credrankCommand: Command = async (args, std) => {
...
+  printCredSummaryTable(credGraph);
+
+  taskReporter.finish("credrank");
   return 0;
 };
 
-/** Find addresses of all nodes matching any of the scoring prefixes. */
-function findScoringAddresses(
+function printCredSummaryTable(credGraph: CredGraph) {
+  console.log(`# Top Participants By Cred`);
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/cli/credrank.js:97]: Done.

[the remark this answers is not in the document]
Comment (decentralion) [reply src/cli/credrank.js:97]: I can see you really wanted this one ;-)

File: src/core/markovProcessGraph.js [hunk 3]
```diff
@@ -104,6 +105,12 @@ export type MarkovEdge = {|
   // to 1.0 for a given `src`.
   +transitionProbability: TransitionProbability,
 |};
+
+export type Participant = {|
```
[the remark this answers is not in the document]
Comment (decentralion) [reply src/core/markovProcessGraph.js:110]: Think of this as a placeholder for just directly using identities, or something similar. We're gonna want to include the identity id here, and regular nodes don't have them.

File: src/core/credGraph.js [hunk 4]
```diff
@@ -70,9 +79,28 @@ export class CredGraph {
...
+      const epochs = this._mpg.epochBoundaries().map((epochStart) => ({
+        type: "USER_EPOCH",
+        owner: address,
+        epochStart,
+      }));
+      let totalCred = 0;
+      const credPerEpoch = epochs.map((e) => {
+        const payoutEdgeAddress = payoutAddressForEpoch(e);
+        const payoutMarkovEdgeAddress = MarkovEdgeAddress.fromParts([
+          "F",
```
Comment (wchargin) [inline src/core/credGraph.js:93]: Prior to this patch, the only occurrence of the `"B"`/`"F"` constants
was in the `rawEdgeAddress` function. I’m wary about duplicating it
here. What do you think about pulling out something like

```javascript
function edgeToMarkovEdge(
  edgeAddress: EdgeAddressT,
  direction: "B" /* Backward */ | "F" /* Forward */
): MarkovEdgeAddressT {
  return MarkovEdgeAddress.fromParts([
    direction,
    ...EdgeAddress.toParts(edge.address),
  ]);
}

function rawEdgeAddress(edge: MarkovEdge): MarkovEdgeAddressT {
  return edgeToMarkovEdge(edge.reversed ? "B" : "F", edge.address);
}
```

and then calling `edgeToMarkovEdge` directly here?

**EDIT:** Just noticed this is in `CredGraph`. In that case, I double
down: the `B`/`F` encoding is 100% an implementation detail of the MPG
module. OK to export this putative `edgeToMarkovEdge` function.

File: src/core/markovProcessGraph.js [hunk 5]
```diff
@@ -668,7 +695,11 @@ export class MarkovProcessGraph {
     return toCompat(COMPAT_INFO, {
       nodes: MapUtil.toObject(this._nodes),
       edges: MapUtil.toObject(this._edges),
-      scoringAddresses: Array.from(this._scoringAddresses),
+      participants: this._participants,
+      finiteEpochBoundaries: this._epochBoundaries.slice(
+        1,
+        this._epochBoundaries.length - 1
```
Comment (wchargin) [inline src/core/markovProcessGraph.js:701]: `.slice(1, -1)`

**Review by wchargin — APPROVED** [review]
LGTM, but please do address inlines or discuss with me first.

**Pull request merged**